index.d.ts

```diff
@@ -0,0 +1,140 @@
+import * as EventEmitter from "eventemitter3";
+
+type Parameter = { [key: string]: string | number | null };
+
+export interface Renderer<TemplateInputType, ElementType, TemplateType> {
+  render(
+    context: {
+      template: TemplateInputType;
+      element: ElementType;
+      content: any;
+      parameters: Parameter;
+    },
+    callback: (err: any, output: TemplateType) => void
+  ): void;
+  reset(
+    context: {
+      domApi: TemplateType;
+      content: any;
+      template: TemplateInputType;
+      parameters: Parameter;
+    },
+    callback: (err: any, output?: TemplateType) => void
+  ): void;
+  destroy(
+    domApi: TemplateType,
+    callback: () => void
+  ): void;
+  getChildElement(
+    domApi: TemplateType,
+    cb: (err: any, element: ElementType) => void
+  ): void;
+}
+
+export interface State {
+  name: string;
+  parameters: Parameter;
+}
+
+export interface RouterOptions {
+  pathPrefix?: string;
+  router?: any; // FIXME
+  throwOnError?: boolean;
+}
+
+export interface StateResolveCallback<ContentType> {
+  (err: any, content?: ContentType): void;
+  redirect(stateName: string, parameters?: Parameter): void;
+}
+
+export interface StateOption<
+  TemplateInputType,
+  TemplateType,
+  DataType,
+  ContentType
+> {
+  name: string;
+  template: TemplateInputType;
+  route?: string;
+  defaultChild?: string | (() => string);
+  data?: DataType;
+  querystringParameters?: string[];
+  defaultParameters?: Parameter;
+  defaultQuerystringParameters?: Parameter;
+
+  resolve?(
+    data: DataType,
+    parameters: Parameter,
+    callback: StateResolveCallback<ContentType>
+  ): void | Promise<never>;
```

Comment:
`asr` supports returning values from `state.resolve`.  See https://github.com/tehshrike/abstract-state-router#returning-values for the docs about this.

```suggestion
  ): void | Promise<never> | ContentType | Promise<ContentType>;
```